11_notify/willshen8/cmd/puppy-server/main.go

```diff
@@ -0,0 +1,107 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+
+	"github.com/anz-bank/go-course/11_notify/willshen8/pkg/puppy"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+var out io.Writer = os.Stdout
+
+var (
+	args     = os.Args[1:]
+	fileName = kingpin.Flag("data", "Puppy Json File Name").Short('d').File()
+	port     = kingpin.Flag("port", "Port of the server").Short('p').Default("9999").Int()
+	storer   = kingpin.Flag("store", "Map/Sync Store").Short('s').Default("map").Enum("map", "sync")
+)
+
+// Config contains arguments parsed from commandline flags.
+type Config struct {
+	file   *os.File
+	port   int
+	storer string
+}
+
+func main() {
+	cmdConfigs, err := parseCmdArgs(args)
+	if err != nil {
+		fmt.Println(err)
+	}
+	newStorer, storerErr := createStorer(cmdConfigs.storer)
+	if storerErr != nil {
+		panic(storerErr)
+	}
+	decodedPuppies, parseErr := parseJSONPuppies(cmdConfigs.file)
+	if parseErr != nil {
+		panic(parseErr)
+	}
+	portFlag, portErr := parsePortFlag(cmdConfigs)
+	if portErr != nil {
+		fmt.Println(portErr)
+	}
+	port := ":" + portFlag
+
+	puppyStorer, _ := createPuppies(newStorer, decodedPuppies)
+	fmt.Fprintln(out, puppyStorer)
+	puppyHandler := puppy.NewRestHandler(puppyStorer)
+	router := puppy.SetupRouter()
+	puppy.SetupRoutes(router, *puppyHandler)
+	log.Fatal(http.ListenAndServe(port, router))
+}
+
+// creatStorer will take the commandline arg and create the appropriate store
+func createStorer(store string) (puppy.Storer, error) {
+	switch store {
+	case "map":
+		return puppy.NewMapStore(), nil
+	case "sync":
+		return puppy.NewSyncStore(), nil
+	}
+	return nil, errors.New("map/sync are the only acceptable flag values")
+}
+
+func parseCmdArgs(args []string) (Config, error) {
+	var parsedConig Config
+	if _, parseError := kingpin.CommandLine.Parse(args); parseError != nil {
+		return parsedConig, parseError
+	}
+	return Config{file: *fileName, port: *port, storer: *storer}, nil
+}
+
+// parseJSONPuppies read from io input and unmarshal them into puyppy JSONS
+func parseJSONPuppies(r io.Reader) ([]puppy.Puppy, error) {
+	var puppies []puppy.Puppy
+	jsonData, _ := ioutil.ReadAll(r) // error already handled by kingpin
+	if err := json.Unmarshal(jsonData, &puppies); err != nil {
+		return nil, err
+	}
+	return puppies, nil
+}
+
+// parsePortFlag check for valid port number entered from commandLine
+func parsePortFlag(c Config) (string, error) {
+	if c.port < 0 || c.port > 65535 {
+		return "0", errors.New("invalid port number entered")
+	}
+	return strconv.Itoa(c.port), nil
+}
+
+// createPuppies takes an array of puppies and saves it to a puppy storer
+func createPuppies(s puppy.Storer, puppies []puppy.Puppy) (puppy.Storer, error) {
+	for _, jsonPuppy := range puppies {
+		jsonPuppy := jsonPuppy // fixes linter issue
```

Comment:
I struggled with this one, but sure what is complaining. So have to just silence it here.

Reply:
It is complaining about that you are taking a reference to the loop variable and passing that reference to a function. It is described in more detail here: https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables

but the gist of it is that there is only one variable for the loop variable and it changes each iteration. Because you are calling `CreatePuppy` through an interface, there is no way to know what it is doing with that reference. If it stores it or passes it to a closure or goroutine, then the value may change on the next loop iteration.

Your solution was correct - `jsonPuppy := jsonPuppy` - this creates a new variable for each iteration of the loop, so on the next iteration, the variable will not be overwritten.

My comment was with regards to the comment `// fixes linter issue`. Our goal is not to fix linter issues, our goal is to write correct code. When the linter points out an issue, we should examine what the linter is saying, and if it is correct, change the code so that our code is correct. If the linter is being too cautious and the code is correct, then we can add a `//nolint:scopelint` comment to tell the linter that the code is ok.

When I said `//nolint:<linter>` I didn't mean that exact string. You're meant to replace `<linter>` with the name of the linter that is reporting the problem (`scopelint` in this case).

So the best way to change your code is to put `jsonPuppy := jsonPuppy` back and remove the comments. Correct code needs no comments (unless it is doing something really tricky, but this idiom inside a loop in Go is quite common and most Go coders will learn to recognise what is happening there).

Reply:
Another link to read: https://www.evanjones.ca/go-gotcha-loop-variables.html, or just search for [golang loop variable reference] to see a lot more that has been written about this.